Declare inputs and outputs for package and packageMobile tasks.

wenslayer opened the pull request:
This is a simple approach to declaring inputs and outputs for the packaging tasks (to address #173). This is intended to start a conversation (@SlevinBE?), as it could likely stand some improvement. Currently, I think it may suffer from:
* It only declares the .swf file as an input -- a good start, but it could also pull in any additional file that goes into the package?
* It feels like there's some code duplication with addMainSwf(), but that function is run during execution, so further refactoring would be needed to use the same bit of code.
* I tried declaring initInputOutputFiles() as a private void method, but Gradle did not like that, so a def it is.
* This approach, if reasonable, could also be applied to the publish task.

Thanks for any feedback.

wenslayer commented on src/main/groovy/org/gradlefx/tasks/AirPackage.groovy:
```diff
@@ -65,6 +69,18 @@ class AirPackage extends DefaultTask {
+        }
+
+        inputs.files swfFileName
+        outputs.files project.file(project.buildDir.name + '/' + flexConvention.output).absolutePath
```
I realized I mixed some conventions in constructing the filenames, so it could be made more consistent.

wenslayer commented:
My most recent commit adds more files (keystore, provisioningProfile, applicationDescriptor) to the set of inputs, and does so in a more consistent way.

SlevinBE commented on src/main/groovy/org/gradlefx/tasks/AirPackage.groovy:
```diff
@@ -65,6 +69,32 @@ class AirPackage extends DefaultTask {
     }
 
+    def initInputOutputFiles() {
+        if (project.type == FlexType.air) {
```
you don't need this check. This task is only used for AIR builds.

wenslayer replied:
I thought so too. In fact, I didn't have this initially, but I noticed that when I had logging statements in here, they were being displayed for every project regardless of type (including swf, swc, and swcAir). Hence this if statement. 

Perhaps that's harmless though? Or maybe there's a better way to do this?

SlevinBE commented on src/main/groovy/org/gradlefx/tasks/AirPackage.groovy:
```diff
@@ -65,6 +69,32 @@ class AirPackage extends DefaultTask {
+            if (flexConvention.air.keystore) {
+                def keystore = project.file(flexConvention.air.keystore)
+                if (keystore.exists()) {
+                    inputs.files keystore.absolutePath
```
I think that, by calling inputs.files multiple times, you're overriding the setting instead of adding something to it. So it probably needs to be set in one go.

wenslayer replied:
I found this works, for I had logging statements at the end of this closure to iterate through the inputs and outputs to display them, and each file declared in this manner was displayed.

Note that no assignment (=) is used.

SlevinBE commented:
This seems like a reasonable approach, straightforward while still being clean.
I agree that the addMainSwf can be factored out, so that both AirPackage and BaseAirMobilePackage can use it, since they use exactly the same code. But I would favor composition over inheritance, since it will become messy quite quickly otherwise (I've seen it happen in the cli code, which I've refactored because of this)

wenslayer commented:
Understood, thanks for reviewing!

I should add that I'm using this locally to build our 130+ project build, with 5 package[Mobile] targets at the end of the dependency chain, and rerunning builds with no changes works as expected for those packaging targets (by stating they're UP-TO-DATE). 

In one of my packaging projects, I have it such that the airMobile.target property can be changed through a Gradle command line property (-Ptarget=foo), so I manually that as an input to the task. It's conceivable that it could go here, but I wonder if that would be over-reaching in a way. Thoughts?

SlevinBE commented:
It solves a quite specific use-case in your build, but I wouldn't add it to GradleFx.
Pretty cool build you've got going there. Wasn't even aware GradleFx was used in such complex projects :)

wenslayer commented:
Thanks for your feedback, I agree, it's a specific use case, so I'll keep it in my gradle file instead of putting it in the plugin.

wenslayer commented:
At this point, I believe I've addressed your feedback, @SlevinBE. If you're satisfied with the PR, would you kindly consider accepting it? I'd be happy to continue discussion on any of your outstanding concerns.

SlevinBE commented:
Thanks for the PR! It has been merged and is now available on the snapshot repo

wenslayer commented:
Great, thanks! BTW, I see it available as [1.2.0-SNAPSHOT](https://oss.sonatype.org/content/repositories/snapshots/org/gradlefx/gradlefx/1.2.0-SNAPSHOT/). Should the version not be 1.2.**1**-SNAPSHOT?

SlevinBE commented:
Indeed, I haven't upstepped the version yet since the latest release. Will do this asap
